src/util/bson/bson.cpp

```diff
@@ -0,0 +1,1170 @@
+            }
+            return *this;
+        }
+
+        BsonContainer(BsonContainer&& v) {
+            m_type = v.m_type;
+
+            if (m_type == DOCUMENT) {
+                document = v.document;
+                v.document = nullptr;
+            } else {
+                array = v.array;
+                v.array = nullptr;
+            }
+        }
+
+        bool is_array() const { return m_type == ARRAY; }
+        bool is_document() const { return m_type == DOCUMENT; }
+
+        void push_back(std::pair<std::string, Bson> value) {
+            if (m_type == DOCUMENT) {
+                (*document)[value.first] = value.second;
+            } else {
+                array->emplace_back(value.second);
+            }
+        }
+
+        std::pair<std::string, Bson> back()
+        {
+            REALM_ASSERT(size() > 0);
+            if (m_type == DOCUMENT) {
+                auto pair = document->back();
+                return pair;
+            } else {
+                return {"", array->back()};
+            }
+        }
+
+        void pop_back()
+        {
+            if (m_type == DOCUMENT) {
+                document->pop_back();
+            } else {
+                array->pop_back();
+            }
+        }
+
+        size_t size() const
+        {
+            if (m_type == DOCUMENT) {
+                return document->size();
+            } else {
+                return array->size();
+            }
+        }
+    };
+
+    std::stack<BsonContainer> m_marks;
+    std::stack<Instruction> m_instructions;
+};
+
+struct BsonError : public std::runtime_error {
+    BsonError(std::string message) : std::runtime_error(message)
+    {
+    }
+};
+
+static constexpr const char * key_number_int                 = "$numberInt";
+static constexpr const char * key_number_long                = "$numberLong";
+static constexpr const char * key_number_double              = "$numberDouble";
+static constexpr const char * key_number_decimal             = "$numberDecimal";
+static constexpr const char * key_timestamp                  = "$timestamp";
+static constexpr const char * key_timestamp_time             = "t";
+static constexpr const char * key_timestamp_increment        = "i";
+static constexpr const char * key_date                       = "$date";
+static constexpr const char * key_object_id                  = "$oid";
+static constexpr const char * key_max_key                    = "$maxKey";
+static constexpr const char * key_min_key                    = "$minKey";
+static constexpr const char * key_regular_expression         = "$regularExpression";
+static constexpr const char * key_regular_expression_pattern = "pattern";
+static constexpr const char * key_regular_expression_options = "options";
+static constexpr const char * key_binary                     = "$binary";
+static constexpr const char * key_binary_base64              = "base64";
+static constexpr const char * key_binary_sub_type            = "subType";
```

Comment:
It might be nice to make these public in a header somewhere so that they can be reused, at least I noticed that the "app" code is hardcoding these strings everywhere too. Maybe a future improvement though (non-blocking).

Reply:
Hmm, the "app" code shouldn't be doing that. I'll look into it in a future PR.